File: worker/sloptic_web_worker/platform.py
"""Platform hosts, read from the list the web half also reads.

CLAUDE.md makes the platform floor STRUCTURAL: someone on team.vercel.app cannot publish
_sloptic.team.vercel.app because the zone is not theirs, so the second factor is unavailable by
construction and no amount of good faith produces it.

The API refuses these when a claim is created, which is where a person is told why. This module is
the same refusal at grade time, and it is not redundant: a grant can predate a suffix being added to
the list, or be written by hand, and may_grade_actively is the last place anyone asks whether attack
traffic may be sent at an app.
"""
from __future__ import annotations

import json
import pathlib

_LIST = pathlib.Path(__file__).resolve().parents[2] / "shared" / "platform-suffixes.json"


def _load() -> tuple[str, ...]:
    try:
        return tuple(json.loads(_LIST.read_text())["suffixes"])
    except Exception:  # noqa: BLE001
        # Fails CLOSED on the entries that matter most. A missing or unreadable list must not turn
        # into "no host is a platform host", which would be the permissive reading of a broken file.
        return ("vercel.app", "netlify.app", "github.io", "pages.dev", "workers.dev",
                "herokuapp.com", "onrender.com", "fly.dev", "web.app", "firebaseapp.com")


SUFFIXES = _load()


def platform_suffix(host: str) -> str | None:
    """The platform suffix this host sits under, or None when it looks like its own domain."""
    h = (host or "").lower().rstrip(".")
    for suffix in SUFFIXES:
        if h == suffix or h.endswith("." + suffix):
            return suffix
    return None

File: web/lib/platform.ts
/** Hosts whose DNS zone belongs to the platform, not to the person deploying on it.
 *
 *  This is not a blocklist and it is not about trust. Owner verification needs TWO independent
 *  proofs, and the second one is a TXT record in the zone. Somebody on `team.vercel.app` cannot
 *  publish under `_sloptic.team.vercel.app` because they do not hold that zone, so the factor is
 *  structurally unavailable to them: no amount of good faith makes it obtainable. Offering the flow
 *  anyway would mean asking for something impossible and then blaming them for not producing it.
 *
 *  Their path to an active grade is a custom domain, which is exactly what verifying is FOR, so the
 *  refusal names it.
 *
 *  CLAUDE.md is explicit that we never substitute a weaker same-surface proof here: two files at two
 *  paths would be one factor written twice.
 */
const PLATFORM_SUFFIXES = [
  "vercel.app",
  "netlify.app",
  "netlify.com",
  "github.io",
  "gitlab.io",
  "pages.dev",
  "workers.dev",
  "herokuapp.com",
  "onrender.com",
  "fly.dev",
  "railway.app",
  "up.railway.app",
  "streamlit.app",
  "replit.app",
  "repl.co",
  "glitch.me",
  "surge.sh",
  "firebaseapp.com",
  "web.app",
  "appspot.com",
  "azurewebsites.net",
  "cloudfront.net",
  "amplifyapp.com",
  "deno.dev",
  "val.run",
  "ngrok.io",
  "ngrok-free.app",
  "trycloudflare.com",
  "loca.lt",
  "vercel.sh",
];

/** The platform suffix this host sits under, or null when the host looks like its own domain.
 *
 *  An exact match counts too: `vercel.app` itself is not somebody's site. */
export function platformSuffix(host: string): string | null {
  const h = host.toLowerCase().replace(/\.$/, "");
  for (const suffix of PLATFORM_SUFFIXES) {
    if (h === suffix || h.endsWith(`.${suffix}`)) return suffix;
  }
  return null;
}

export const PLATFORM_REFUSAL =
  "That is a platform subdomain, so the DNS half of verification is not something you can publish: " +
  "the zone belongs to the platform, not to you. Attach a custom domain and verify that instead. " +
  "The passive checks run on it either way.";
